### PR TITLE
Remove cache in Heads

### PR DESCRIPTION
**Pull request by maximedenes**
This cache makes the pretyper depend on components that should morally
be higher-level (Libobject and co), so I'd like to see how critical this
cache is before taking any action.

### CODE REVIEW

**Comment by maximedenes** [comment]
Benchmark running here: https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/669/console

**Comment by gares** [comment]
Very effective cache indeed

**Comment by maximedenes** [comment]
Benchmark result:

```
┌────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬─────────────────────────────┐
│                        │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │         mem faults          │
│                        │                         │                                             │                                             │                               │                             │
│           package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │     NEW     OLD     PDIFF   │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│          coq-fiat-core │  110.96  113.89 -2.57 % │      315003143719      323653454257 -2.67 % │      392301050535      400667770079 -2.09 % │     505908     511236 -1.04 % │     479     217   +120.74 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│           coq-compcert │ 1035.37 1060.76 -2.39 % │     2887096421890     2956167937041 -2.34 % │     4456535272566     4474279444729 -0.40 % │    1293852    1298748 -0.38 % │     268     235    +14.04 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│       coq-fiat-parsers │  704.92  716.00 -1.55 % │     1972801750029     2003922278046 -1.55 % │     3001234093333     3028601998451 -0.90 % │    3025856    3095544 -2.25 % │     448     847    -47.11 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│              coq-color │  712.06  722.56 -1.45 % │     1988544840898     2020108155507 -1.56 % │     2361695769258     2402266510460 -1.69 % │    1348012    1401716 -3.83 % │     141     226    -37.61 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│            coq-bignums │   72.42   73.36 -1.28 % │      201857970621      203979825495 -1.04 % │      266979946323      268900886390 -0.71 % │     502416     504780 -0.47 % │     176     170     +3.53 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│ coq-mathcomp-ssreflect │   39.97   40.48 -1.26 % │      110449931658      111360785088 -0.82 % │      132112860212      133049086502 -0.70 % │     527088     529920 -0.53 % │      20      17    +17.65 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│             coq-sf-plf │   44.10   44.60 -1.12 % │      123442881832      125397763018 -1.56 % │      158381652447      160528632246 -1.34 % │     495644     495736 -0.02 % │      42       0      +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│             coq-geocoq │ 2024.47 2042.42 -0.88 % │     5649795265021     5700311849705 -0.89 % │     8185647002839     8218574423382 -0.40 % │    1196708    1198400 -0.14 % │     147     272    -45.96 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│         coq-coquelicot │   77.40   78.07 -0.86 % │      212107472561      214795820967 -1.25 % │      255959253625      258795092106 -1.10 % │     699560     701552 -0.28 % │     212      10  +2020.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│  coq-mathcomp-solvable │  201.81  203.51 -0.84 % │      562801683617      567533663731 -0.83 % │      750593601378      755101330708 -0.60 % │     806272     806288 -0.00 % │       8       1   +700.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│     coq-mathcomp-field │  238.59  240.26 -0.70 % │      664526414377      668450697773 -0.59 % │      954177049469      954661676315 -0.05 % │     757860     770192 -1.60 % │       4       1   +300.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│   coq-mathcomp-algebra │  175.66  176.88 -0.69 % │      490138449646      492598974024 -0.50 % │      610617476488      612590648816 -0.32 % │     640944     640324 +0.10 % │      13       7    +85.71 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│              coq-verdi │  127.04  127.86 -0.64 % │      352706469970      355525129873 -0.79 % │      454838862232      457458888136 -0.57 % │     595624     606472 -1.79 % │       8       4   +100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│            coq-unimath │ 2107.20 2120.42 -0.62 % │     5868943358575     5907841016774 -0.66 % │    10247697861171    10259717566886 -0.12 % │     908764     962408 -5.57 % │     490     184   +166.30 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│       coq-math-classes │  219.48  220.62 -0.52 % │      613213157423      617049377282 -0.62 % │      771331189722      777670239992 -0.82 % │     514100     519092 -0.96 % │      91      16   +468.75 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│              coq-flocq │  676.53  679.82 -0.48 % │     1883926450287     1891017494665 -0.37 % │     2693617231847     2690361565249 +0.12 % │    1726316    1734644 -0.48 % │     285     358    -20.39 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│         coq-verdi-raft │ 1407.20 1413.35 -0.44 % │     3926733639589     3938487921820 -0.30 % │     5331784549638     5350630731209 -0.35 % │    1804688    1805940 -0.07 % │       0       2   -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│  coq-mathcomp-fingroup │   55.86   56.10 -0.43 % │      155275835261      156203632326 -0.59 % │      202723332486      203422352366 -0.34 % │     579124     582060 -0.50 % │     126       1 +12500.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│ coq-mathcomp-character │  206.59  207.23 -0.31 % │      575750047522      576917079557 -0.20 % │      766104632966      767119425649 -0.13 % │    1079756    1086300 -0.60 % │       1      40    -97.50 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│        coq-lambda-rust │ 2095.98 2100.76 -0.23 % │     5842617439982     5857595900657 -0.26 % │     8021133857500     8037083573561 -0.20 % │    1395320    1388396 +0.50 % │      54       5   +980.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│               coq-corn │ 1511.02 1512.81 -0.12 % │     4215750064260     4222412916498 -0.16 % │     6193578756654     6223028316699 -0.47 % │     836844     868816 -3.68 % │       1       7    -85.71 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│        coq-fiat-crypto │ 4410.28 4414.46 -0.09 % │    12254629888032    12262265571757 -0.06 % │    19481352452016    19536756699416 -0.28 % │    2488156    2673724 -6.94 % │    1080     961    +12.38 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│ coq-fiat-crypto-legacy │ 6704.34 6706.48 -0.03 % │    18679127931517    18685611129623 -0.03 % │    30156238333701    30195512558576 -0.13 % │    2478836    2489036 -0.41 % │    1175    1203     -2.33 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│               coq-hott │  348.38  346.88 +0.43 % │      946315225266      943733188627 +0.27 % │     1421938249730     1425640216830 -0.26 % │     478088     479580 -0.31 % │     439      41   +970.73 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────────┤
│ coq-mathcomp-odd-order │ 1442.98 1429.06 +0.97 % │     4021457306539     3984066614452 +0.94 % │     6597640725376     6593268575416 +0.07 % │    1352688    1273620 +6.21 % │      27     123    -78.05 % │
└───────────────────
```

So this is ready for merge.

**Comment by maximedenes** [comment]
@mattam82 can you be the assignee?

**Comment by ejgallego** [comment]
A couple of comments, out of curiosity:

- what was the purpose of this cache, potentially it seems useful; was the problem that the miss rate was near 100%: or there was some other reason for its ineffectiveness.
- @maximedenes when you write "the pretyper depend on components that should morally be higher-level (Libobject and co)" I think there is a little catch here, in that even if it works for this; it is gonna be tricky not to allow components to register state; is your plan indeed to make all such state functional?

  Or just let me rephrase the question, what would have been the plan if we'd had find out that the cache was critical.

**Comment by maximedenes** [comment]
> * is your plan indeed to make all such state functional?

Yes. We are not so far from that, in fact, so that's good news!

> * what would have been the plan if we'd had find out that the cache was critical.

No idea, but I'm afraid the problems we actually have give me enough to think about ;)

**Review by gares — APPROVED** [review]

**Comment by ejgallego** [comment]
Note the commit introducing this 37c82d53d56816c1f01062abd20c93e6a22ee924 is quite interesting; it also introduces some other infrastructure that I have moved to the vernac layer, except for a nit in pretyping.

I don't have the time to analyze such commit in depth, but I'd be good to understand what was going on IMHO.

**Comment by SkySkimmer** [comment]
>This cache makes the pretyper depend on components that should morally
be higher-level (Libobject and co), so I'd like to see how critical this
cache is before taking any action.

If it had been in the kernel, it may have been critical in the "critical bug" sense ;) considering this PR fixes https://github.com/coq/coq/issues/11133
